From 2b057c4c70b553f66228ae74aeb2736b6d59cded Mon Sep 17 00:00:00 2001
From: Philip Taron <philip@qumulo.com>
Date: Thu, 19 Mar 2026 17:01:05 -0700
Subject: [PATCH 5/7] Port patch 18: Implement gss_get_mic_iov()

Add IOV-based MIC generation following the pattern of gss_get_mic and
gss_(un)wrap_iov. Implemented for CFX (AES+HMAC-SHA) only; errors on
non-CFX crypto. New vtable entry gm_get_mic_iov added to all mechanism
interfaces (NULL for non-krb5 mechs).
---
 lib/gssapi/gssapi/gssapi.h       |   3 +
 lib/gssapi/gssapi_mech.h         |   8 ++
 lib/gssapi/krb5/cfx.c            | 168 +++++++++++++++++++++++++++++++
 lib/gssapi/krb5/external.c       |   1 +
 lib/gssapi/krb5/get_mic.c        |  20 ++++
 lib/gssapi/libgssapi-exports.def |   1 +
 lib/gssapi/mech/gss_get_mic.c    |  29 ++++++
 lib/gssapi/netlogon/external.c   |   1 +
 lib/gssapi/ntlm/external.c       |   1 +
 lib/gssapi/sanon/external.c      |   1 +
 lib/gssapi/spnego/external.c     |   1 +
 lib/gssapi/version-script.map    |   1 +
 12 files changed, 235 insertions(+)

diff --git a/lib/gssapi/gssapi/gssapi.h b/lib/gssapi/gssapi/gssapi.h
index 726543f5a..56c2549a1 100644
--- a/lib/gssapi/gssapi/gssapi.h
+++ b/lib/gssapi/gssapi/gssapi.h
@@ -952,6 +952,9 @@ gss_decapsulate_token(gss_const_buffer_t /* input_token */,
  * GSS_IOV
  */
 
+GSSAPI_LIB_FUNCTION OM_uint32 GSSAPI_LIB_CALL
+gss_get_mic_iov(OM_uint32 *, gss_const_ctx_id_t, gss_iov_buffer_desc *, int);
+
 GSSAPI_LIB_FUNCTION OM_uint32 GSSAPI_LIB_CALL
 gss_wrap_iov(OM_uint32 *, gss_ctx_id_t, int, gss_qop_t, int *,
 	     gss_iov_buffer_desc *, int);
diff --git a/lib/gssapi/gssapi_mech.h b/lib/gssapi/gssapi_mech.h
index ced27b517..65f0249a8 100644
--- a/lib/gssapi/gssapi_mech.h
+++ b/lib/gssapi/gssapi_mech.h
@@ -109,6 +109,13 @@ typedef OM_uint32 GSSAPI_CALLCONV _gss_get_mic_t
 	       gss_buffer_t            /* message_token */
 	      );
 
+typedef OM_uint32 GSSAPI_CALLCONV _gss_get_mic_iov_t
+	      (OM_uint32 *minor_status,
+	        gss_const_ctx_id_t context_handle,
+	        gss_iov_buffer_desc *iov, /* iov buffers */
+	        int iov_count
+	      );
+
 typedef OM_uint32 GSSAPI_CALLCONV _gss_verify_mic_t
 	      (OM_uint32 *,            /* minor_status */
 	       gss_const_ctx_id_t,     /* context_handle */
@@ -579,6 +586,7 @@ typedef struct gssapi_mech_interface_desc {
 	_gss_delete_sec_context_t	*gm_delete_sec_context;
 	_gss_context_time_t		*gm_context_time;
 	_gss_get_mic_t			*gm_get_mic;
+	_gss_get_mic_iov_t		*gm_get_mic_iov;
 	_gss_verify_mic_t		*gm_verify_mic;
 	_gss_wrap_t			*gm_wrap;
 	_gss_unwrap_t			*gm_unwrap;
diff --git a/lib/gssapi/krb5/cfx.c b/lib/gssapi/krb5/cfx.c
index af50292c2..d67ad97fc 100644
--- a/lib/gssapi/krb5/cfx.c
+++ b/lib/gssapi/krb5/cfx.c
@@ -1791,3 +1791,171 @@ OM_uint32 _gssapi_verify_mic_cfx(OM_uint32 *minor_status,
 
     return GSS_S_COMPLETE | seq_err;
 }
+
+OM_uint32
+_gssapi_get_mic_iov(
+	OM_uint32 *minor_status,
+	gsskrb5_ctx ctx,
+	krb5_context context,
+	gss_iov_buffer_desc *iov,
+	int iov_count)
+{
+	OM_uint32 major_status, junk;
+	gss_iov_buffer_desc *header, *trailer, *padding;
+	size_t gsshsize;
+	size_t gsstsize, k5tsize;
+	int i;
+	gss_cfx_mic_token token;
+	krb5_error_code ret;
+	int32_t seq_number;
+	unsigned usage;
+	krb5_crypto_iov *data = NULL;
+
+	header = _gk_find_buffer(iov, iov_count, GSS_IOV_BUFFER_TYPE_HEADER);
+	if (header == NULL) {
+		*minor_status = EINVAL;
+		return GSS_S_FAILURE;
+	}
+
+	padding = _gk_find_buffer(iov, iov_count, GSS_IOV_BUFFER_TYPE_PADDING);
+	if (padding != NULL) {
+		padding->buffer.length = 0;
+	}
+
+	trailer = _gk_find_buffer(iov, iov_count, GSS_IOV_BUFFER_TYPE_TRAILER);
+
+	major_status = _gk_verify_buffers(
+		minor_status, ctx, header, padding, trailer, FALSE);
+	if (major_status != GSS_S_COMPLETE) {
+		return major_status;
+	}
+
+	*minor_status = krb5_crypto_length(
+		context, ctx->crypto, KRB5_CRYPTO_TYPE_CHECKSUM, &k5tsize);
+	if (*minor_status) {
+		return GSS_S_FAILURE;
+	}
+
+	gsshsize = sizeof(gss_cfx_mic_token_desc);
+	gsstsize = k5tsize;
+
+	if (trailer == NULL) {
+		gsshsize += gsstsize;
+	} else if (GSS_IOV_BUFFER_FLAGS(trailer->type) & GSS_IOV_BUFFER_FLAG_ALLOCATE) {
+		major_status = _gk_allocate_buffer(minor_status, trailer, gsstsize);
+		if (major_status) {
+			goto failure;
+		}
+	} else if (trailer->buffer.length < gsstsize) {
+		*minor_status = KRB5_BAD_MSIZE;
+		major_status = GSS_S_FAILURE;
+		goto failure;
+	} else {
+		trailer->buffer.length = gsstsize;
+	}
+
+	if (GSS_IOV_BUFFER_FLAGS(header->type) & GSS_IOV_BUFFER_FLAG_ALLOCATE) {
+		major_status = _gk_allocate_buffer(minor_status, header, gsshsize);
+		if (major_status != GSS_S_COMPLETE) {
+			goto failure;
+		}
+	} else if (header->buffer.length < gsshsize) {
+		*minor_status = KRB5_BAD_MSIZE;
+		major_status = GSS_S_FAILURE;
+		goto failure;
+	} else {
+		header->buffer.length = gsshsize;
+	}
+
+	token = (gss_cfx_mic_token)header->buffer.value;
+
+	token->TOK_ID[0] = 0x04;
+	token->TOK_ID[1] = 0x04;
+	token->Flags = 0;
+	memset(token->Filler, 0xFF, 5);
+
+	if ((ctx->more_flags & LOCAL) == 0) {
+		token->Flags |= CFXSentByAcceptor;
+	}
+
+	if (ctx->more_flags & ACCEPTOR_SUBKEY) {
+		token->Flags |= CFXAcceptorSubkey;
+	}
+
+	if (ctx->more_flags & LOCAL) {
+		usage = KRB5_KU_USAGE_INITIATOR_SIGN;
+	} else {
+		usage = KRB5_KU_USAGE_ACCEPTOR_SIGN;
+	}
+
+	HEIMDAL_MUTEX_lock(&ctx->ctx_id_mutex);
+	krb5_auth_con_getlocalseqnumber(context, ctx->auth_context, &seq_number);
+	_gss_mg_encode_be_uint32(0, &token->SND_SEQ[0]);
+	_gss_mg_encode_be_uint32(seq_number, &token->SND_SEQ[4]);
+	krb5_auth_con_setlocalseqnumber(context, ctx->auth_context, ++seq_number);
+	HEIMDAL_MUTEX_unlock(&ctx->ctx_id_mutex);
+
+	data = calloc(iov_count + 3, sizeof(data[0]));
+	if (data == NULL) {
+		*minor_status = ENOMEM;
+		major_status = GSS_S_FAILURE;
+		goto failure;
+	}
+
+	/*
+	 * Packet looks like:
+	 * {data | "header" | gss-trailer (krb5 checksum)}
+	 */
+
+	for (i = 0; i < iov_count; i++) {
+		switch (GSS_IOV_BUFFER_TYPE(iov[i].type)) {
+		case GSS_IOV_BUFFER_TYPE_DATA:
+			data[i].flags = KRB5_CRYPTO_TYPE_DATA;
+			break;
+		case GSS_IOV_BUFFER_TYPE_SIGN_ONLY:
+			data[i].flags = KRB5_CRYPTO_TYPE_SIGN_ONLY;
+			break;
+		default:
+			data[i].flags = KRB5_CRYPTO_TYPE_EMPTY;
+			break;
+		}
+		data[i].data.length = iov[i].buffer.length;
+		data[i].data.data = iov[i].buffer.value;
+	}
+
+	data[i].flags = KRB5_CRYPTO_TYPE_DATA;
+	data[i].data.data = header->buffer.value;
+	data[i].data.length = sizeof(gss_cfx_mic_token_desc);
+	i++;
+
+	data[i].flags = KRB5_CRYPTO_TYPE_CHECKSUM;
+	if (trailer) {
+		data[i].data.data = trailer->buffer.value;
+	} else {
+		data[i].data.data =
+			(uint8_t *)header->buffer.value + sizeof(gss_cfx_mic_token_desc);
+	}
+	data[i].data.length = k5tsize;
+	i++;
+
+	ret = krb5_create_checksum_iov(context, ctx->crypto, usage, data, i, NULL);
+	if (ret) {
+		*minor_status = ret;
+		major_status = GSS_S_FAILURE;
+		goto failure;
+	}
+
+	free(data);
+
+	*minor_status = 0;
+	return GSS_S_COMPLETE;
+
+failure:
+	if (data) {
+		free(data);
+	}
+
+	gss_release_iov_buffer(&junk, iov, iov_count);
+
+	return major_status;
+}
diff --git a/lib/gssapi/krb5/external.c b/lib/gssapi/krb5/external.c
index e58df18c5..aa9c82adb 100644
--- a/lib/gssapi/krb5/external.c
+++ b/lib/gssapi/krb5/external.c
@@ -346,6 +346,7 @@ static gssapi_mech_interface_desc krb5_mech = {
     _gsskrb5_delete_sec_context,
     _gsskrb5_context_time,
     _gsskrb5_get_mic,
+    _gsskrb5_get_mic_iov,
     _gsskrb5_verify_mic,
     _gsskrb5_wrap,
     _gsskrb5_unwrap,
diff --git a/lib/gssapi/krb5/get_mic.c b/lib/gssapi/krb5/get_mic.c
index d9cf9d700..e7342d42f 100644
--- a/lib/gssapi/krb5/get_mic.c
+++ b/lib/gssapi/krb5/get_mic.c
@@ -328,3 +328,23 @@ OM_uint32 GSSAPI_CALLCONV _gsskrb5_get_mic
   krb5_free_keyblock (context, key);
   return ret;
 }
+
+OM_uint32 GSSAPI_CALLCONV
+_gsskrb5_get_mic_iov(
+	OM_uint32 *minor_status,
+	gss_const_ctx_id_t context_handle,
+	gss_iov_buffer_desc *iov,
+	int iov_count)
+{
+	krb5_context context;
+	const gsskrb5_ctx ctx = (const gsskrb5_ctx)context_handle;
+
+	GSSAPI_KRB5_INIT(&context);
+
+	if (ctx->more_flags & IS_CFX) {
+		return _gssapi_get_mic_iov(minor_status, ctx, context, iov, iov_count);
+	} else {
+		*minor_status = 0;
+		return GSS_S_FAILURE;
+	}
+}
diff --git a/lib/gssapi/libgssapi-exports.def b/lib/gssapi/libgssapi-exports.def
index 0b4c3949d..5368573d1 100644
--- a/lib/gssapi/libgssapi-exports.def
+++ b/lib/gssapi/libgssapi-exports.def
@@ -43,6 +43,7 @@ EXPORTS
 	gss_export_name_composite
 	gss_export_sec_context
 	gss_get_mic
+	gss_get_mic_iov
 	gss_get_neg_mechs
 	gss_get_name_attribute
         gss_import_cred
diff --git a/lib/gssapi/mech/gss_get_mic.c b/lib/gssapi/mech/gss_get_mic.c
index 8663053d4..934f74710 100644
--- a/lib/gssapi/mech/gss_get_mic.c
+++ b/lib/gssapi/mech/gss_get_mic.c
@@ -49,3 +49,32 @@ gss_get_mic(OM_uint32 *minor_status,
 	return (m->gm_get_mic(minor_status, ctx->gc_ctx, qop_req,
 		    message_buffer, message_token));
 }
+
+GSSAPI_LIB_FUNCTION OM_uint32 GSSAPI_LIB_CALL
+gss_get_mic_iov(
+	OM_uint32 *minor_status,
+	gss_const_ctx_id_t context_handle,
+	gss_iov_buffer_desc *iov,
+	int iov_count)
+{
+	struct _gss_context *ctx = (struct _gss_context *)context_handle;
+	gssapi_mech_interface m;
+
+	if (minor_status) {
+		*minor_status = 0;
+	}
+	if (ctx == NULL) {
+		return GSS_S_NO_CONTEXT;
+	}
+	if (iov == NULL && iov_count != 0) {
+		return GSS_S_CALL_INACCESSIBLE_READ;
+	}
+
+	m = ctx->gc_mech;
+
+	if (m->gm_get_mic_iov == NULL) {
+		return GSS_S_UNAVAILABLE;
+	}
+
+	return (m->gm_get_mic_iov)(minor_status, ctx->gc_ctx, iov, iov_count);
+}
diff --git a/lib/gssapi/netlogon/external.c b/lib/gssapi/netlogon/external.c
index 14f471e8d..41aec54ea 100644
--- a/lib/gssapi/netlogon/external.c
+++ b/lib/gssapi/netlogon/external.c
@@ -48,6 +48,7 @@ static gssapi_mech_interface_desc netlogon_mech = {
     _netlogon_delete_sec_context,
     _netlogon_context_time,
     _netlogon_get_mic,
+    NULL, /* gm_get_mic_iov */
     _netlogon_verify_mic,
     NULL,
     NULL,
diff --git a/lib/gssapi/ntlm/external.c b/lib/gssapi/ntlm/external.c
index ff2cd2d1f..d2da4d488 100644
--- a/lib/gssapi/ntlm/external.c
+++ b/lib/gssapi/ntlm/external.c
@@ -73,6 +73,7 @@ static gssapi_mech_interface_desc ntlm_mech = {
     _gss_ntlm_delete_sec_context,
     _gss_ntlm_context_time,
     _gss_ntlm_get_mic,
+    NULL, /* gm_get_mic_iov */
     _gss_ntlm_verify_mic,
     _gss_ntlm_wrap,
     _gss_ntlm_unwrap,
diff --git a/lib/gssapi/sanon/external.c b/lib/gssapi/sanon/external.c
index 8812f9e08..65f328746 100644
--- a/lib/gssapi/sanon/external.c
+++ b/lib/gssapi/sanon/external.c
@@ -203,6 +203,7 @@ static gssapi_mech_interface_desc sanon_mech = {
     _gss_sanon_delete_sec_context,
     _gss_sanon_context_time,
     _gss_sanon_get_mic,
+    NULL, /* gm_get_mic_iov */
     _gss_sanon_verify_mic,
     _gss_sanon_wrap,
     _gss_sanon_unwrap,
diff --git a/lib/gssapi/spnego/external.c b/lib/gssapi/spnego/external.c
index 2a5121efa..ef33a10bd 100644
--- a/lib/gssapi/spnego/external.c
+++ b/lib/gssapi/spnego/external.c
@@ -97,6 +97,7 @@ static gssapi_mech_interface_desc spnego_mech = {
     _gss_spnego_delete_sec_context,
     _gss_spnego_context_time,
     _gss_spnego_get_mic,
+    NULL, /* gm_get_mic_iov */
     _gss_spnego_verify_mic,
     _gss_spnego_wrap,
     _gss_spnego_unwrap,
diff --git a/lib/gssapi/version-script.map b/lib/gssapi/version-script.map
index f9a24f20c..95ffbd483 100644
--- a/lib/gssapi/version-script.map
+++ b/lib/gssapi/version-script.map
@@ -46,6 +46,7 @@ HEIMDAL_GSS_2.0 {
 		gss_export_name_composite;
 		gss_export_sec_context;
 		gss_get_mic;
+		gss_get_mic_iov;
 		gss_get_neg_mechs;
 		gss_get_name_attribute;
 		gss_import_cred;
-- 
2.53.0

